rust/src/initramfs.rs

```diff
@@ -30,6 +30,22 @@ fn list_files_recurse<P: glib::IsA<gio::Cancellable>>(
     if let Some(c) = cancellable {
         let _ = c.set_error_if_cancelled()?;
     }
+
+    // We need to include all the directories leading up to the path because cpio doesn't do that,
+    // and the kernel at extraction time won't create missing leading dirs. But this also serves
+    // as validating that the path is canonical and doesn't do e.g. /etc/foo/../bar.
+    let mut path_rebuilt = PathBuf::new();
+    for component in Path::new(path).components() {
+        match component {
+            Component::Normal(c) => {
+                path_rebuilt.push(c);
+                // safe to unwrap here since it's all built up of components of a known UTF-8 path
+                filelist.insert(path_rebuilt.to_str().unwrap().into());
```

Comment:
To avoid duplication I think we could keep a cache of the last parent path emitted.  That would fix `--track /etc/foo/bar.conf /etc/foo/baz.conf` to ensure there's just one `/etc/foo`.

Reply:
`filelist` here is a BTreeSet, which already performs deduplication on insertion.